From b32c53e7d2fcc106c13ebfca4870af44423bf7c6 Mon Sep 17 00:00:00 2001
From: google@wwsnet.net <google@wwsnet.net@e8e0d7a0-c8d9-11dd-a880-a1081c7ac358>
Date: Mon, 13 Sep 2010 00:59:11 +0000
Subject: [PATCH] - split the loading part of opkg_conf_init() out into opkg_conf_load()
 - move the temporary destination list into the global config struct

git-svn-id: http://opkg.googlecode.com/svn/trunk@561 e8e0d7a0-c8d9-11dd-a880-a1081c7ac358
---
 libopkg/opkg.c      |    3 +++
 libopkg/opkg_conf.c |   44 ++++++++++++++++++++++++--------------------
 libopkg/opkg_conf.h |    2 ++
 3 files changed, 29 insertions(+), 20 deletions(-)

diff --git a/libopkg/opkg.c b/libopkg/opkg.c
index 87c3244..b36dc1b 100644
--- a/libopkg/opkg.c
+++ b/libopkg/opkg.c
@@ -120,6 +120,9 @@ opkg_new()
 	if (opkg_conf_init())
 		goto err0;
 
+	if (opkg_conf_load())
+		goto err0;
+
 	if (pkg_hash_load_feeds())
 		goto err1;
 
diff --git a/libopkg/opkg_conf.c b/libopkg/opkg_conf.c
index 6dee5ac..22c99eb 100644
--- a/libopkg/opkg_conf.c
+++ b/libopkg/opkg_conf.c
@@ -90,15 +90,15 @@ opkg_option_t options[] = {
 };
 
 static int
-resolve_pkg_dest_list(nv_pair_list_t *nv_pair_list)
+resolve_pkg_dest_list(void)
 {
      nv_pair_list_elt_t *iter;
      nv_pair_t *nv_pair;
      pkg_dest_t *dest;
      char *root_dir;
 
-     for (iter = nv_pair_list_first(nv_pair_list); iter;
-		     iter = nv_pair_list_next(nv_pair_list, iter)) {
+     for (iter = nv_pair_list_first(&conf->tmp_dest_list); iter;
+		     iter = nv_pair_list_next(&conf->tmp_dest_list, iter)) {
 	  nv_pair = (nv_pair_t *)iter->data;
 
 	  if (conf->offline_root) {
@@ -184,8 +184,7 @@ opkg_conf_set_option(const char *name, const char *value)
 
 static int
 opkg_conf_parse_file(const char *filename,
-				pkg_src_list_t *pkg_src_list,
-				nv_pair_list_t *tmp_dest_nv_pair_list)
+				pkg_src_list_t *pkg_src_list)
 {
      int line_num = 0;
      int err = 0;
@@ -268,7 +267,7 @@ opkg_conf_parse_file(const char *filename,
 				regmatch[11].rm_eo - regmatch[11].rm_so);
 	  }
 
-	  /* We use the tmp_dest_nv_pair_list below instead of
+	  /* We use the conf->tmp_dest_list below instead of
 	     conf->pkg_dest_list because we might encounter an
 	     offline_root option later and that would invalidate the
 	     directories we would have computed in
@@ -291,7 +290,7 @@ opkg_conf_parse_file(const char *filename,
 				   "Skipping.\n", name, value);
 	       }
 	  } else if (strcmp(type, "dest") == 0) {
-	       nv_pair_list_append(tmp_dest_nv_pair_list, name, value);
+	       nv_pair_list_append(&conf->tmp_dest_list, name, value);
 	  } else if (strcmp(type, "lists_dir") == 0) {
 	       conf->lists_dir = xstrdup(value);
 	  } else if (strcmp(type, "arch") == 0) {
@@ -411,9 +410,19 @@ glob_errfunc(const char *epath, int eerrno)
 int
 opkg_conf_init(void)
 {
+	pkg_src_list_init(&conf->pkg_src_list);
+	pkg_dest_list_init(&conf->pkg_dest_list);
+	pkg_dest_list_init(&conf->tmp_dest_list);
+	nv_pair_list_init(&conf->arch_list);
+
+	return 0;
+}
+
+int
+opkg_conf_load(void)
+{
 	int i, glob_ret;
 	char *tmp, *tmp_dir_base, **tmp_val;
-	nv_pair_list_t tmp_dest_nv_pair_list;
 	glob_t globbuf;
 	char *etc_opkg_conf_pattern;
 
@@ -423,11 +432,6 @@ opkg_conf_init(void)
 	conf->check_x509_path = 1;
 #endif
 
-	pkg_src_list_init(&conf->pkg_src_list);
-	pkg_dest_list_init(&conf->pkg_dest_list);
-	nv_pair_list_init(&conf->arch_list);
-	nv_pair_list_init(&tmp_dest_nv_pair_list);
-
 	if (!conf->offline_root)
 		conf->offline_root = xstrdup(getenv("OFFLINE_ROOT"));
 
@@ -438,7 +442,7 @@ opkg_conf_init(void)
 			goto err0;
 		}
 		if (opkg_conf_parse_file(conf->conf_file,
-				&conf->pkg_src_list, &tmp_dest_nv_pair_list))
+				&conf->pkg_src_list))
 			goto err1;
 	}
 
@@ -467,7 +471,7 @@ opkg_conf_init(void)
 					!strcmp(conf->conf_file, globbuf.gl_pathv[i]))
 				continue;
 		if ( opkg_conf_parse_file(globbuf.gl_pathv[i],
-			&conf->pkg_src_list, &tmp_dest_nv_pair_list)<0) {
+			&conf->pkg_src_list)<0) {
 			globfree(&globbuf);
 			goto err1;
 		}
@@ -531,16 +535,16 @@ opkg_conf_init(void)
 	}
 
 	/* Even if there is no conf file, we'll need at least one dest. */
-	if (nv_pair_list_empty(&tmp_dest_nv_pair_list)) {
-	nv_pair_list_append(&tmp_dest_nv_pair_list,
+	if (nv_pair_list_empty(&conf->tmp_dest_list)) {
+		nv_pair_list_append(&conf->tmp_dest_list,
 			OPKG_CONF_DEFAULT_DEST_NAME,
 			OPKG_CONF_DEFAULT_DEST_ROOT_DIR);
 	}
 
-	if (resolve_pkg_dest_list(&tmp_dest_nv_pair_list))
+	if (resolve_pkg_dest_list())
 		goto err4;
 
-	nv_pair_list_deinit(&tmp_dest_nv_pair_list);
+	nv_pair_list_deinit(&conf->tmp_dest_list);
 
 	return 0;
 
@@ -580,7 +584,7 @@ err1:
 		}
 	}
 err0:
-	nv_pair_list_deinit(&tmp_dest_nv_pair_list);
+	nv_pair_list_deinit(&conf->tmp_dest_list);
 	if (conf->dest_str)
 		free(conf->dest_str);
 	if (conf->conf_file)
diff --git a/libopkg/opkg_conf.h b/libopkg/opkg_conf.h
index 7496ff4..0b163b0 100644
--- a/libopkg/opkg_conf.h
+++ b/libopkg/opkg_conf.h
@@ -46,6 +46,7 @@ struct opkg_conf
 {
      pkg_src_list_t pkg_src_list;
      pkg_dest_list_t pkg_dest_list;
+     pkg_dest_list_t tmp_dest_list;
      nv_pair_list_t arch_list;
 
      int restrict_to_default_dest;
@@ -132,6 +133,7 @@ struct opkg_option {
 };
 
 int opkg_conf_init(void);
+int opkg_conf_load(void);
 void opkg_conf_deinit(void);
 
 int opkg_conf_write_status_files(void);
-- 
1.7.1

